File: app/models/note.rb
# frozen_string_literal: true

class Note < ApplicationRecord
  FREE_COUNT_LIMIT = 100

  include CarrierWave::Base64ImageContent
  include PgSearch::Model

  belongs_to :user

  pg_search_scope(
    :search_by_title_and_content,
    against: {
      title: 'A',
      text_content: 'B'
    },
    using: {
      tsearch: {
        prefix: true,
        tsvector_column: %i[tsv_title tsv_content]
      }
    }
  )

  has_paper_trail skip: %i[tsv_title tsv_content]

  mount_uploaders :images, ImageUploader

  base64_image_content content: :text_content, images: :images

  scope :default_ordered, -> { order(updated_at: :desc) }

  scope :archived, -> { where.not(archived_at: nil) }
  scope :unarchived, -> { where(archived_at: nil) }

  scope :having_todo_tasks, -> { where(%(tasks @> '[{ "done": false }]')) }
  scope :having_done_tasks, -> { where(%(tasks @> '[{ "done": true }]')) }
  scope :having_tasks, -> { having_todo_tasks.or(having_done_tasks) }

  validate(
    :does_not_exceed_free_count_limit,
    on: :create,
    if: ->(note) { note.user.free? }
  )

  before_save :populate_tasks

  def to_param
    uid
  end

  def as_json(options = {})
    only = options.fetch(
      :only,
      %i[uid title created_at updated_at archived_at content tasks]
    ).map(&:to_s)
    options = { only: only }.merge(options)

    # use the actual content (see `HasContent`)
    json = super(options).merge('content' => content)

    # we need to slice again for `content` to be considered as well
    json.slice(*only)
  end

  def dup
    super.tap do |dup_note|
      dup_note.uid = SecureRandom.uuid

      dup_note.images = images.map do |image|
        File.open(image.file.file)
      end
    end
  end

  def archive!
    update!(archived_at: Time.zone.now)
  end

  private

  def does_not_exceed_free_count_limit
    return if user.notes.count < FREE_COUNT_LIMIT

    errors.add(
      :base,
      "You have reached your note limit of #{FREE_COUNT_LIMIT} notes. " \
      'Please delete some notes or upgrade your subscription.'
    )
  end

  def populate_tasks
    NoteTaskPopulator.new(self).run
  end
end

File: client2/src/Note.js
import Uuid from './Uuid';

class Note {
  constructor(uid, title, content, createdAt, updatedAt, archivedAt) {
    this._uid = uid || Uuid.generateV4();
    this._title = title || '';
    this._content = content || '';
    this._createdAt = createdAt || new Date();
    this._updatedAt = updatedAt || this._createdAt;
    this._serverUpdatedAt = this._updatedAt;
    this._archivedAt = archivedAt;
  }

  get uid() { return this._uid }

  get title() { return this._title }
  set title(title) {
    this._updatedAt = new Date();
    this._title = title;
  }

  get content() { return this._content }
  set content(content) {
    this._updatedAt = new Date();
    this._content = content;
  }

  get createdAt() { return this._createdAt }
  get updatedAt() { return this._updatedAt }

  get serverUpdatedAt() { return this._serverUpdatedAt }
  set serverUpdatedAt(serverUpdatedAt) { this._serverUpdatedAt = serverUpdatedAt }

  get archivedAt() { return this._archivedAt; }

  setArchived() {
    this._archivedAt = new Date();
  }

  isNew() {
    return !this.title && !this.content;
  }

  toJson() {
    return JSON.stringify(
      {
        uid: this.uid,
        title: this.title,
        content: this.content,
        created_at: this.createdAt.toISOString(),
        updated_at: this.updatedAt.toISOString(),
        server_updated_at: this.serverUpdatedAt.toISOString(),
        archived_at: this.archivedAt ? this.archivedAt.toISOString() : null
      }
    );
  }

  static fromAttributes(attributes) {
    return new Note(
      attributes.uid,
      attributes.title,
      attributes.content,
      new Date(attributes.created_at),
      new Date(attributes.updated_at),
      attributes.archived_at ? new Date(attributes.archived_at) : null
    );
  }
}

export default Note;
